python_modules/libraries/dagster-aws/dagster_aws/emr/emr_eks_pyspark_step_launcher.py

```diff
@@ -0,0 +1,569 @@
+    This works by making the resource also inherit from `StepLauncher`,
+    and implement `launch_step` to serialize the definition of the step,
+    run it on the remote cluster.
+
+    Dagster operations produce events (e.g., when a step finishes, or when
+    an output is produced). The events produced by the operation running
+    on EMR on EKS in the Spark driver thus need to be serialized, sent
+    back to the Dagster instance, and deserialized so that Dagster is
+    aware of what happened in the remote cluster. Dagster events are not
+    JSON serialized, and have to be pickled. As these events are small,
+    the following procedure is used:
+
+    - In the Spark driver, as part of the job entrypoint (`job_main.py`),
+      events produced by the step are pickled, converted to Base64, and
+      logged to `stdout`.
+
+    - Logs of the Spark driver are automatically reported by EMR to a
+      CloudWatch log group.
+
+    - The step launcher running on the original Dagster instance tails
+      the CloudWatch logs. Logs are reported to Dagster. When an event
+      is seen in a log record, it is decoded, unpickled, and sent to
+      Dagster.
+
+    Logs are uploaded to CloudWatch by EMR every 10s. As such, events may
+    not be reported to Dagster in real time, and there can be some delays.
+    Events will also often show up in Dagster in batches, when a batch
+    of new logs has been pulled. Although more involved solutions might
+    allow to reduce latency (e.g., creating per-job run AWS SQS queues),
+    using the logs should be good enough for now.
+
+    Deploying Code
+
+    The resource supports two ways of deploying our Python code to the
+    EMR on EKS jobs:
+
+    - **Running a container that contains the code.** This is the target
+      workflow for a cloud deployment.
+
+      In this case, all of the required code is baked in the image.
+
+      This workflow is not amenable to local development, though.
+
+    - **Running a container without the code, uploading the code at job
+      submission**. This is the target workflow for local development.
+
+      When submitting a job, a ZIP archive of the code is created and
+      uploaded to S3. The Spark driver will then pull the ZIP archive,
+      add it to the Python path, and use that to import modules.
+
+      **Note**: Python implicit namespace packages are not supported.
+      All packages **must** have an `__init__.py` to be importable
+      in this workflow.
+
+      **Note**: if the container used has some code, it will just be
+      ignored. For instance, we can run with a container that has the code
+      from `develop` or a previous version of the code from the branch. As
+      we are staging the local code, code on the container will be ignored.
+
+    The container-only workflow is selected automatically when the Dagster
+    repository is seen to come from a docker image. In other cases, the
+    code-upload workflow is used.
```

Comment:
nit: I think this section is helpful, but we could indicate the corresponding config parameters to use for each case

Reply:
I added some references